src/rapids_singlecell/preprocessing/_normalize.py

```diff
@@ -98,14 +97,113 @@ def normalize_total(
+    mul_kernel(
+        (math.ceil(X.shape[0] / 128),),
+        (128,),
+        (X.indptr, X.data, X.shape[0], int(target_sum)),
+    )
+    return X
+
+
+def _normalize_total_dask(X: DaskArray, target_sum: int) -> DaskArray:
+    if isinstance(X._meta, sparse.csr_matrix):
+        from ._kernels._norm_kernel import _mul_csr
+
+        mul_kernel = _mul_csr(X.dtype)
+        mul_kernel.compile()
+
+        def __mul(X_part):
+            mul_kernel(
+                (math.ceil(X_part.shape[0] / 32),),
+                (32,),
+                (X_part.indptr, X_part.data, X_part.shape[0], int(target_sum)),
+            )
+            return X_part
+
+        X = X.map_blocks(__mul, meta=_meta_sparse(X.dtype))
+    elif isinstance(X._meta, cp.ndarray):
+        from ._kernels._norm_kernel import _mul_dense
+
+        mul_kernel = _mul_dense(X.dtype)
+        mul_kernel.compile()
+
+        def __mul(X_part):
+            mul_kernel(
+                (math.ceil(X_part.shape[0] / 128),),
+                (128,),
+                (X_part, X_part.shape[0], X_part.shape[1], int(target_sum)),
+            )
+            return X_part
+
+        X = X.map_blocks(__mul, meta=_meta_dense(X.dtype))
+    else:
+        raise ValueError(f"Cannot normalize {type(X)}")
+    return X
+
+
+def _get_target_sum(X) -> int:
+    if isinstance(X, sparse.csr_matrix):
+        return _get_target_sum_csr(X)
+    elif isinstance(X, DaskArray):
+        return _get_target_sum_dask(X)
+    else:
+        return cp.median(X.sum(axis=1))
+
+
+def _get_target_sum_csr(X: sparse.csr_matrix) -> int:
+    from ._kernels._norm_kernel import _get_sparse_sum_major
+
+    counts_per_cell = cp.zeros(X.shape[0], dtype=X.dtype)
+    sum_kernel = _get_sparse_sum_major(X.dtype)
+    sum_kernel(
+        (X.shape[0],),
+        (64,),
+        (X.indptr, X.data, counts_per_cell, X.shape[0]),
+    )
+
+    target_sum = cp.median(counts_per_cell)
+    return target_sum
+
+
+def _get_target_sum_dask(X: DaskArray) -> int:
+    if isinstance(X._meta, sparse.csr_matrix):
+        from ._kernels._norm_kernel import _get_sparse_sum_major
+
+        sum_kernel = _get_sparse_sum_major(X.dtype)
+        sum_kernel.compile()
+
+        def __sum(X_part):
+            counts_per_cell = cp.zeros(X_part.shape[0], dtype=X_part.dtype)
+            sum_kernel(
+                (X.shape[0],),
+                (64,),
+                (X_part.indptr, X_part.data, counts_per_cell, X_part.shape[0]),
+            )
+            return counts_per_cell
+
+    elif isinstance(X._meta, cp.ndarray):
+
+        def __sum(X_part):
+            return X_part.sum(axis=1)
+    else:
+        raise ValueError(f"Cannot compute target sum for {type(X)}")
+    target_sum_chunk_matrices = X.map_blocks(
+        __sum,
+        meta=cp.array((1.0,), dtype=X.dtype),
+        dtype=X.dtype,
+        chunks=(X.chunksize[0],),
+        drop_axis=1,
+    )
+    counts_per_cell = target_sum_chunk_matrices.compute()
```

Comment:
I believe we have an implementation in scanpy that makes this lazy as well: 

https://github.com/scverse/scanpy/blob/be99b230fa84e077f5167979bc9f6dacc4ad0d41/src/scanpy/preprocessing/_normalization.py#L34-L48

probably worth trying out since this is fairly expensive (i.e., requires a full-pass over the data)

Reply:
I'll investigate if that a solution for me aswell

Reply:
I'll keep this for now

Reply:
This faster and scales better